### PR TITLE
Better support for enum trees

### PR DESCRIPTION
**Pull request by soc**
Replace the approach of “detect some magic combination of flags to
trigger some action” by introducing an enum flag which makes the
semantics and the intentions of the code using it more explicit.

This basically involves wiring up the existing ACC_ENUM bit to a new
modifier flag and making sure it is set correctly when encountering
enums.

The existing enum tests files/pos/t5165 and files/pos/t2764 keep
working, showing that this hasn't introduced any obvious regressions.

Tests for the changes in Namer which prevent scalac from messing with
enum trees can't be added yet, because one currently can't define an
enum in Scala without the help of the macro paradise.
The intention is to add the @enum macro as a full test suite as soon
as one can depend on macro annotations.

### CODE REVIEW

**Comment by retronym** [comment]
We'll need to find a way to test this. One idea: a partest-case can provide a compiler plugin, see `test/files/run/analyzerPlugins.scala`.

I presume you'll have to duplicate some effort in the GenBCode backend.

The commit message should lay out the motivation for doing this, too.

**Comment by soc** [comment]
Yes, agreed. My intention was to get the basic stuff for further work running and to submit the enum macro as a test case when everything runs fine.

I'll update GenBCode and expand the commit message.

**Comment by soc** [comment]
Could you have a look again, @retronym? Thanks!

**Comment by adriaanm** [comment]
Please rebase on latest master. Some PR validation improvements I did ran into problems fixed by #3114.

**Comment by soc** [comment]
Is there anything still missing?

**Comment by retronym** [comment]
- [ ] test cases still missing
- [ ] the use of a flag (which are a scarce resource) should be discussed on scala-internals.

**Comment by soc** [comment]
- There are two existing test cases already which seem to do their job (I discovered them when they started failing while working on this). Is there anything else missing which wouldn't be tested best by a full blown @enum macro?
- Will do. EDIT: Done: https://groups.google.com/forum/#!searchin/scala-internals/enum/scala-internals/uvifzou5RQc/zmHdoijkCFwJ

**Comment by retronym** [comment]
Please mention and review those existing test cases in the commit comment.

**Comment by retronym** [comment]
Your message to scala-internals doesn't seem to mention this PR nor the addition of a new flag. Please take your time and explain this so that the readers can make an informed decision without trawling through big threads.

**Comment by retronym** [comment]
Okay, I see you sent a [second message](https://groups.google.com/forum/#!topic/scala-internals/BKNrZ2FwATM) that does mention the flags.

**Comment by soc** [comment]
The commit message reads

```
“[...] The existing enum tests files/pos/t5165 and files/pos/t2764 keep
working, showing that this hasn't introduced any obvious regressions.”
```

I think they cover basic (t2764) and advanced uses of enums (t5165 ... defining multiple enums inside an annotation and using these types as annotation members again ... I think it can't get any messier :-D).

**Comment by JamesIry** [comment]
If the consensus is to go ahead and use a flag I'd like to throw my vote in to defer this to 2.12. We're very late in the 2.11 cycle and I would rather spend time and more importantly risk on fixing user visible issues than cleaning up internals.

**Comment by soc** [comment]
I can live without the new flag, but having the rest of the fixes ship with 2.11 would be nice, because without it we would have yet another release where one can't even test/develop enum functionality without hacking the compiler. Which means development gets pushed back, therefore testing gets pushed back, the SIP gets pushed back and rolling it out to end users gets pushed back. That's one thing I'd like to avoid.

File: src/reflect/scala/reflect/internal/Flags.scala [hunk 1]
```diff
@@ -119,6 +119,7 @@ class ModifierFlags {
   final val DEFAULTINIT   = 1L << 41      // symbol is initialized to the default value: used by -Xcheckinit
   final val ARTIFACT      = 1L << 46      // symbol should be ignored when typechecking; will be marked ACC_SYNTHETIC in bytecode
   final val DEFAULTMETHOD = 1L << 47      // symbol is a java default method
+  final val ENUM          = 1L << 48      // symbol is an enum
```
Comment (gkossakowski) [inline src/reflect/scala/reflect/internal/Flags.scala:122]: I think we could use a little bit more documentation. Is it only for symbols derived from Java sources/class files?
Is this applied to class symbol itself or maybe also to it's companion class? Or maybe it's also applied to accessors for each enum case?

I know that other flags do not offer more than you offer but we want to rise the bar higher than in the past.

**Comment by adriaanm** [comment]
If there's a clear path to improving Java interop, we should consider this for M8. Cleanups are great, but if they don't fix any issues for the 2.11 release, they should go to 2.12, as James said. We should probably branch 2.12 soon, but that's also work to get our integration infrastructure set up (it shouldn't be, but it's pretty time consuming)./

**Comment by soc** [comment]
Can we get a decision on this whether it goes into 2.11 or 2.12? Personally, I'll follow the branch into which this is merged anyway, but it would be kind of sad for people who want to test `@enum` stuff if they couldn't do it for the next 18 moths without switching to trunk.
@gkossakowski I'll add documentation.

**Comment by retronym** [comment]
Maybe I'm missing this, but I don't see any description of behavioural changes as a result of the patches to `Namers`. Are there any? If not, I would only expect to see changes in code that used to check for a flag-combo to check for proposed `ENUM` flag. 

In particular, these places where changed:

``` scala
    def noEnterGetterSetter(vd: ValDef) = !vd.mods.isLazy && (
         !owner.isClass
      || (vd.mods.isPrivateLocal && !vd.mods.isCaseAccessor)
      || (vd.name startsWith nme.OUTER)
      || (context.unit.isJava)
      || isEnumItem(vd)
    )

    def noFinishGetterSetter(vd: ValDef) = (
         (vd.mods.isPrivateLocal && !vd.mods.isLazy) // all lazy vals need accessors, even private[this]
      || vd.symbol.isModuleVar
      || isEnumItem(vd))
```

**Comment by retronym** [comment]
"Cleanup enum support in the backend" is really quite misleading; we've never had such a thing :)

So, if the backend(s) has been expanded to emit enums, we need really need a test. That will requires a macro and can't be expressed in sources, a fact that should be clearly described in the commit message.

**Comment by soc** [comment]
@retronym The changes in `Namers` exist to prevent scalac from messing with enums which don't come from Java, e. g. ones defined in Scala (by whatever means). An alternative would be to set the Java flag when creating enums, but that would run against the idea of this commit of cleaning up the non-obvious use of flag combos.

I'll adjust the commit message accordingly.

I'll also adjust the naming of `isEnumElement`. Both the JLS and the JVMS talk about “enum constants”, I guess we should just follow that naming.

**Comment by soc** [comment]
I guess I should also find a better name for `isEnum`, I'd use `isEnumClass` although I'm concerned that it would mistakenly considered to be a check for `== java.lang.Enum`.

**Comment by soc** [comment]
@retronym, @gkossakowski Updated commit message, added documentation, cleaned up code, moved `isEnumClass` from Symbol to Namer.

**Comment by JamesIry** [comment]
ouch

**Comment by gkossakowski** [comment]
Why this test broke? Why positions are different?

**Comment by soc** [comment]
@gkossakowski It seems to be a common theme for that the last few days that things work fine locally but PR validation keeps failing.

**Comment by soc** [comment]
PLS REBUILD/pr-scala@7693133

**Comment by soc** [comment]
Review by @retronym, please!

**Comment by adriaanm** [comment]
> @gkossakowski It seems to be a common theme for that the last few days that things work fine locally but PR validation keeps failing.

that definitely warrants investigation!

**Comment by soc** [comment]
Not ready for prime time. :-/
I just found an issue resulting in “illegal cyclic reference involving class Day”: https://gist.github.com/soc/7620539
I'm currently investigating.

**Comment by soc** [comment]
Ok, found the issue with @xeno-by's help. Looks like calling `superClass` is pretty dangerous.

**Pull request closed without merging**